### PR TITLE
framework/eventloop: Modify eventloop timer APIs

### PR DESCRIPTION
**Pull request by jeongarmy**
- framework/eventloop: Add data size parameter to eventloop_add_timer/eventloop_add_timer_async
- framework/eventloop: Change return type of timeout_callback and logic about cleaning used resources

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [ed4f73ca21d44eea7af85f80ae36a4b9bda10c4e] -

**Comment by seinfra** [comment]
Target : [ed4f73ca21d44eea7af85f80ae36a4b9bda10c4e] -

**Comment by seinfra** [comment]
Target : [7e35c13636cc1932b88f6dc7345d0d931de8c651] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7e35c13636cc1932b88f6dc7345d0d931de8c651] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c5e15ad3a16af4f955da134f19d8c75ccf77e474] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c5e15ad3a16af4f955da134f19d8c75ccf77e474] - Code Rule Check OK.

File: framework/src/eventloop/eventloop_timer.c [hunk 1]
```diff
@@ -76,9 +174,27 @@ static el_timer_t *add_timer(el_loop_t *loop, unsigned int timeout, bool repeat,
...
+			EL_FREE(timer);
+			return NULL;
+		}
+		memcpy(callback->cb_data, data, data_size);
+	}
 	timer->data = (void *)callback;
 
+	ret = eventloop_register_timer(timer);
+	if (ret != OK) {
+		EL_FREE(callback);
```
Comment (jeongchanKim) [inline framework/src/eventloop/eventloop_timer.c]: Is EL_FREE(callback->cb_data) not needed?

File: framework/src/eventloop/eventloop_timer.c [hunk 2]
```diff
@@ -76,9 +174,27 @@ static el_timer_t *add_timer(el_loop_t *loop, unsigned int timeout, bool repeat,
...
+		callback->cb_data = NULL;
+	} else {
+		callback->cb_data = EL_ALLOC(data_size);
+		if (callback->cb_data == NULL) {
+			eldbg("Failed to allocate callback data\n");
+			EL_FREE(callback);
+			EL_FREE(timer);
+			return NULL;
+		}
+		memcpy(callback->cb_data, data, data_size);
```
Comment (jeongchanKim) [inline framework/src/eventloop/eventloop_timer.c]: for memcpy, string.h is needed.

File: framework/src/eventloop/eventloop_timer.c [hunk 3]
```diff
@@ -34,11 +35,95 @@ struct timer_cb_s {
 };
 typedef struct timer_cb_s timer_cb_t;
 
+/* The stucture for wrapping of timer handle to be kept in a list internally. */
```
Comment (jeongchanKim) [inline framework/src/eventloop/eventloop_timer.c]: stucture -> structure

File: framework/src/eventloop/eventloop_timer.c (hunk 1 above)
Comment (jeongarmy) [reply framework/src/eventloop/eventloop_timer.c]: Done

**Comment by seinfra** [comment]
Target : [437b744a498001626e9b5b23bfccffd84440308a] - Code Rule Check OK.

File: framework/src/eventloop/eventloop_timer.c [hunk 4]
```diff
@@ -113,24 +237,29 @@ int eventloop_delete_timer(el_timer_t *timer)
 		return EVENTLOOP_INVALID_PARAM;
 	}
 
-	/* Stop the timer */
-	uv_timer_stop(timer);
+	if (!is_registered_timer(timer) || uv__is_closing(timer)) {
```
Comment (tyha1115) [inline framework/src/eventloop/eventloop_timer.c]: uv__is_closing is declared at uv__handle.h.
include uv_handle.h is needed.

**Pull request closed without merging**